src/storage/txn/flow_controller.rs

```diff
@@ -0,0 +1,934 @@
+// Copyright 2021 TiKV Project Authors. Licensed under Apache-2.0.
+
+use std::collections::VecDeque;
+use std::f64::INFINITY;
+use std::sync::atomic::{AtomicU64, Ordering};
+use std::sync::mpsc::{self, Receiver, RecvTimeoutError, Sender};
+use std::sync::Arc;
+use std::thread::{Builder, JoinHandle};
+use std::time::Duration;
+use std::u64;
+
+use collections::HashMap;
+use engine_rocks::FlowInfo;
+use engine_traits::{CFNamesExt, MiscExt};
+use rand::Rng;
+use tikv_util::time::{duration_to_sec, Consume, Instant, Limiter};
+
+use crate::storage::config::Config;
+use crate::storage::metrics::*;
+use crate::storage::Engine;
+
+const ADJUST_DURATION: u64 = 1000; // 1000ms
+const RATIO_PRECISION: f64 = 10000000.0;
+const EMA_FACTOR: f64 = 0.6;
+const LIMIT_UP_PERCENT: f64 = 0.04; // 4%
+const LIMIT_DOWN_PERCENT: f64 = 0.02; // 2%
+const MIN_THROTTLE_SPEED: f64 = 16.0 * 1024.0; // 16KB
+const MAX_THROTTLE_SPEED: f64 = 200.0 * 1024.0 * 1024.0; // 200MB
+
+#[derive(Eq, PartialEq)]
+enum Trend {
+    Increasing,
+    Decreasing,
+    NoTrend,
+    OnlyOne,
+}
+
+// Flow controller is used to throttle the write rate at scheduler level, aiming
+// to substitute the write stall mechanism of RocksDB. It features in two points:
+//   * throttle at scheduler, so raftstore and apply won't be blocked anymore
+//   * better control on the throttle rate to avoid QPS drop under heavy write
+//
+// When write stall happens, the max speed of write rate max_delayed_write_rate
+// is limited to 16MB/s by default which doesn't take real disk ability into
+// account. It may underestimate the disk's throughout that 16MB/s is too small
+// at once, causing a very large jitter on the write duration.
+// Also, it decreases the delayed write rate further if the factors still exceed
+// the threshold. So under heavy write load, the write rate may be throttled to
+// a very low rate from time to time, causing QPS drop eventually.
+//
+// This main idea of the flow controller is to throttle at a steady write rate
+// so that the number of L0 keeps around the threshold. When it falls below the
+// threshold, the throttle state wouldn't exit right away. Instead, it may keep
+// or increase the throttle speed depending on some statistics.
+// How can we decide the throttle speed? As we can imagine, the consumption
+// ability of L0 wouldn't change dramatically corresponding to the ability of
+// hardware. So we can record the flush flow when reach the threshold as target
+// flow, and increase or decrease the throttle speed based on whether current
+// flush flow is smaller/larger than target flow.
+
+// For compaction pending bytes, we use discardable ratio...(TBD)
+pub struct FlowController {
+    discard_ratio: Arc<AtomicU64>,
+    limiter: Arc<Limiter>,
+    tx: Sender<bool>,
```

Comment:
Why is this needed? I thought `flow_receiver` will return error if server is shutdown.

Reply:
For dynamical configurable, it's needed